pipeline/pipeline.go

```diff
@@ -81,18 +81,69 @@ func (r *Runtime) Run() error {
 	return r.err
 }
 
-//
-//
-//
+// Updates the current status of a step
+func (r *Runtime) traceStep(processState *backend.State, err error, step *backend.Step) error {
+	if r.tracer == nil {
+		// no tracer nothing to trace :)
+		return nil
+	}
 
-func (r *Runtime) execAll(procs []*backend.Step) <-chan error {
+	if processState == nil {
+		processState = new(backend.State)
+		if err != nil {
+			processState.Exited = true
+			processState.OOMKilled = false
+			processState.ExitCode = 126 // command invoked cannot be executed.
+		}
+	}
+
+	state := new(State)
+	state.Pipeline.Time = r.started
+	state.Pipeline.Step = step
+	state.Process = processState // empty
+	state.Pipeline.Error = r.err
+
+	traceErr := r.tracer.Trace(state)
+	if traceErr != nil {
+		return traceErr
+	}
+	return err
+}
+
+// Executes a set of parallel steps
+func (r *Runtime) execAll(steps []*backend.Step) <-chan error {
 	var g errgroup.Group
 	done := make(chan error)
 
-	for _, proc := range procs {
-		proc := proc
+	for _, step := range steps {
+		step := step
```

Comment:
`step := step` 🤔

Reply:
Yea :)

Reply:
WE do need it ... it seems go replaces the value of step from the loop.

Reply:
yes because the step var is used in a go call afterwards ...